File: solutions/lengthOfLongestSubstring-3/lengthOfLongestSubstring-3.cpp

/**********************************************************************************
*
* Given a string s, find the length of the longest substring without repeating characters.
*
* Input: s = "abcabcbb"
* Output: 3
*
*
**********************************************************************************/

class Solution {

  /*
  *   [Sliding Window]
  *
  *   Time complexity : O(2n)=O(n). In the worst case each character will be visited twice by i and j.
  *
  *   Space complexity : O(min(m,n)). Same as the previous approach.
  *   We need O(k) space for the sliding window, where k is the size of the Set.
  *   The size of the Set is upper bounded by the size of the string n and the size of the charset/alphabet m.
  */

public:
    int lengthOfLongestSubstring(string s) {

        unordered_map<char, int> window;
        int n = s.length();

        int left = 0;
        int right = 0;
        int valid = 0;
        char c;
        int result = 0;
        while (right < n) {

            c = s[right];
            right++;
            window[c]++;
            while (window[c] > 1) {

                window[s[left]]--;
                left++;
            }
            result = max(result, right-left);
        }
        return result;
    }
};

File: solutions/lengthOfLongestSubstring-3/lengthOfLongestSubstring-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        n = len(s)
        if n == 0:
            return 0
        window = {}
        start = 0
        result = 1

        for i in range(n):
            c = s[i]
            if c in window:
                start = max(start, window[c]+1)
            result = max(result, i-start+1)
            window[c] = i
        return result
